File: lib/testy_mcp/tools/bugs.rb
# frozen_string_literal: true

module TestyMcp
  module Tools
    class ListBugs < MCP::Tool
      tool_name "list_bugs"
      description "List bugs with optional filters for status, tags, date range, and search."
      annotations(read_only_hint: true, destructive_hint: false)

      input_schema(
        properties: {
          status: { type: "string", enum: %w[open resolved], description: "Filter by status" },
          feature_tag: { type: "string", description: "Filter by feature tag" },
          cause_tag: { type: "string", description: "Filter by cause tag" },
          date_from: { type: "string", description: "Filter from date (YYYY-MM-DD)" },
          date_until: { type: "string", description: "Filter until date (YYYY-MM-DD)" },
          search: { type: "string", description: "Search by title" },
          page: { type: "integer", description: "Page number for pagination" }
        },
      )

      class << self
        include Base

        def call(status: nil, feature_tag: nil, cause_tag: nil, date_from: nil, date_until: nil, search: nil, page: nil, server_context:)
          client = server_context[:client]
          auth_error = require_auth!(client)
          return auth_error if auth_error

          params = {
            status: status, feature_tag: feature_tag, cause_tag: cause_tag,
            date_from: date_from, date_until: date_until, search: search, page: page
          }.compact
          result = client.get("/api/v1/bugs", params)

          return error_response(result) unless result.success?

          text_response(JSON.pretty_generate(result.body))
        end
      end
    end

    class GetBug < MCP::Tool
      tool_name "get_bug"
      description "Get a bug by ID with full details."
      annotations(read_only_hint: true, destructive_hint: false)

      input_schema(
        properties: {
          id: { type: "integer", description: "Bug ID" }
        },
        required: ["id"]
      )

      class << self
        include Base

        def call(id:, server_context:)
          client = server_context[:client]
          auth_error = require_auth!(client)
          return auth_error if auth_error

          result = client.get("/api/v1/bugs/#{id}")

          return error_response(result) unless result.success?

          text_response(JSON.pretty_generate(result.body))
        end
      end
    end

    class CreateBug < MCP::Tool
      tool_name "create_bug"
      description "Create a new bug report."
      annotations(read_only_hint: false, destructive_hint: false)

      input_schema(
        properties: {
          title: { type: "string", description: "Bug title" },
          description: { type: "string", description: "Bug description" },
          steps_to_reproduce: { type: "string", description: "Steps to reproduce the bug" },
          obtained_result: { type: "string", description: "What actually happened" },
          expected_result: { type: "string", description: "What was expected" },
          feature_tag: { type: "string", description: "Feature tag" },
          cause_tag: { type: "string", description: "Root cause tag" },
          status: { type: "string", enum: %w[open resolved], description: "Bug status (default: open)" }
        },
        required: ["title", "description"]
      )

      class << self
        include Base

        def call(title:, description:, steps_to_reproduce: nil, obtained_result: nil, expected_result: nil, feature_tag: nil, cause_tag: nil, status: nil, server_context:)
          client = server_context[:client]
          auth_error = require_auth!(client)
          return auth_error if auth_error

          body = {
            bug: {
              title: title, description: description, steps_to_reproduce: steps_to_reproduce,
              obtained_result: obtained_result, expected_result: expected_result,
              feature_tag: feature_tag, cause_tag: cause_tag, status: status
            }.compact
          }
          result = client.post("/api/v1/bugs", body)

          return error_response(result) unless result.success?

          text_response(JSON.pretty_generate(result.body))
        end
      end
    end

    class UpdateBug < MCP::Tool
      tool_name "update_bug"
      description "Update an existing bug."
      annotations(read_only_hint: false, destructive_hint: false)

      input_schema(
        properties: {
          id: { type: "integer", description: "Bug ID" },
          title: { type: "string", description: "New title" },
          description: { type: "string", description: "New description" },
          steps_to_reproduce: { type: "string", description: "New steps to reproduce" },
          obtained_result: { type: "string", description: "New obtained result" },
          expected_result: { type: "string", description: "New expected result" },
          feature_tag: { type: "string", description: "New feature tag" },
          cause_tag: { type: "string", description: "New cause tag" },
          status: { type: "string", enum: %w[open resolved], description: "New status" }
        },
        required: ["id"]
      )

      class << self
        include Base

        def call(id:, title: nil, description: nil, steps_to_reproduce: nil, obtained_result: nil, expected_result: nil, feature_tag: nil, cause_tag: nil, status: nil, server_context:)
          client = server_context[:client]
          auth_error = require_auth!(client)
          return auth_error if auth_error

          body = {
            bug: {
              title: title, description: description, steps_to_reproduce: steps_to_reproduce,
              obtained_result: obtained_result, expected_result: expected_result,
              feature_tag: feature_tag, cause_tag: cause_tag, status: status
            }.compact
          }
          result = client.patch("/api/v1/bugs/#{id}", body)

          return error_response(result) unless result.success?

          text_response(JSON.pretty_generate(result.body))
        end
      end
    end

    class DeleteBug < MCP::Tool
      tool_name "delete_bug"
      description "Delete a bug."
      annotations(read_only_hint: false, destructive_hint: true)

      input_schema(
        properties: {
          id: { type: "integer", description: "Bug ID" }
        },
        required: ["id"]
      )

      class << self
        include Base

        def call(id:, server_context:)
          client = server_context[:client]
          auth_error = require_auth!(client)
          return auth_error if auth_error

          result = client.delete("/api/v1/bugs/#{id}")

          return error_response(result) unless result.success?

          text_response("Bug #{id} deleted.")
        end
      end
    end
  end
end

File: cli/internal/commands/bugs.go
package commands

import (
	"encoding/json"
	"fmt"

	"github.com/spf13/cobra"

	"github.com/victorbitancourt/testy-cli/internal/appctx"
	"github.com/victorbitancourt/testy-cli/internal/output"
)

// NewBugsCmd creates the bugs command group.
func NewBugsCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:     "bugs",
		Aliases: []string{"bug"},
		Short:   "Manage bugs",
	}
	cmd.AddCommand(newBugsListCmd())
	cmd.AddCommand(newBugsShowCmd())
	cmd.AddCommand(newBugsCreateCmd())
	cmd.AddCommand(newBugsUpdateCmd())
	cmd.AddCommand(newBugsDeleteCmd())
	return cmd
}

func newBugsListCmd() *cobra.Command {
	var (
		status     string
		featureTag string
		causeTag   string
		search     string
		dateFrom   string
		dateUntil  string
		page       string
	)

	cmd := &cobra.Command{
		Use:   "list",
		Short: "List bugs",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			app := appctx.FromContext(cmd.Context())
			if err := requireAuth(app); err != nil {
				return app.Err(err)
			}

			params := map[string]string{}
			if status != "" {
				params["status"] = status
			}
			if featureTag != "" {
				params["feature_tag"] = featureTag
			}
			if causeTag != "" {
				params["cause_tag"] = causeTag
			}
			if search != "" {
				params["search"] = search
			}
			if dateFrom != "" {
				params["date_from"] = dateFrom
			}
			if dateUntil != "" {
				params["date_until"] = dateUntil
			}
			if page != "" {
				params["page"] = page
			}

			data, err := app.Client.Get("/api/v1/bugs", params)
			if err != nil {
				return app.Err(err)
			}

			var resp struct {
				Bugs []any          `json:"bugs"`
				Meta map[string]any `json:"meta"`
			}
			if err := json.Unmarshal(data, &resp); err != nil {
				return app.Err(fmt.Errorf("failed to parse response: %w", err))
			}

			return app.OK(resp.Bugs,
				output.WithSummary(fmt.Sprintf("%d bug(s)", len(resp.Bugs))),
				output.WithMeta(resp.Meta),
			)
		},
	}
	cmd.Flags().StringVar(&status, "status", "", "Filter by status")
	cmd.Flags().StringVar(&featureTag, "feature-tag", "", "Filter by feature tag")
	cmd.Flags().StringVar(&causeTag, "cause-tag", "", "Filter by cause tag")
	cmd.Flags().StringVar(&search, "search", "", "Search by title/description")
	cmd.Flags().StringVar(&dateFrom, "date-from", "", "Filter from date (YYYY-MM-DD)")
	cmd.Flags().StringVar(&dateUntil, "date-until", "", "Filter until date (YYYY-MM-DD)")
	cmd.Flags().StringVar(&page, "page", "", "Page number")
	return cmd
}

func newBugsShowCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "show <id>",
		Short: "Show a bug",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			app := appctx.FromContext(cmd.Context())
			if err := requireAuth(app); err != nil {
				return app.Err(err)
			}

			data, err := app.Client.Get("/api/v1/bugs/"+args[0], nil)
			if err != nil {
				return app.Err(err)
			}

			var resp struct {
				Bug any `json:"bug"`
			}
			if err := json.Unmarshal(data, &resp); err != nil {
				return app.Err(fmt.Errorf("failed to parse response: %w", err))
			}

			return app.OK(resp.Bug)
		},
	}
}

func newBugsCreateCmd() *cobra.Command {
	var (
		description string
		steps       string
		obtained    string
		expected    string
		featureTag  string
		causeTag    string
		status      string
	)

	cmd := &cobra.Command{
		Use:   "create <title>",
		Short: "Create a bug",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			app := appctx.FromContext(cmd.Context())
			if err := requireAuth(app); err != nil {
				return app.Err(err)
			}

			bug := map[string]any{
				"title":       args[0],
				"description": description,
			}
			if steps != "" {
				bug["steps_to_reproduce"] = steps
			}
			if obtained != "" {
				bug["obtained_result"] = obtained
			}
			if expected != "" {
				bug["expected_result"] = expected
			}
			if featureTag != "" {
				bug["feature_tag"] = featureTag
			}
			if causeTag != "" {
				bug["cause_tag"] = causeTag
			}
			if status != "" {
				bug["status"] = status
			}

			body := map[string]any{"bug": bug}
			data, err := app.Client.Post("/api/v1/bugs", body)
			if err != nil {
				return app.Err(err)
			}

			var resp struct {
				Bug any `json:"bug"`
			}
			if err := json.Unmarshal(data, &resp); err != nil {
				return app.Err(fmt.Errorf("failed to parse response: %w", err))
			}

			return app.OK(resp.Bug, output.WithSummary("Bug created"))
		},
	}
	cmd.Flags().StringVar(&description, "description", "", "Bug description")
	cmd.Flags().StringVar(&steps, "steps", "", "Steps to reproduce")
	cmd.Flags().StringVar(&obtained, "obtained", "", "Obtained result")
	cmd.Flags().StringVar(&expected, "expected", "", "Expected result")
	cmd.Flags().StringVar(&featureTag, "feature-tag", "", "Feature tag")
	cmd.Flags().StringVar(&causeTag, "cause-tag", "", "Root cause tag")
	cmd.Flags().StringVar(&status, "status", "", "Bug status")
	return cmd
}

func newBugsUpdateCmd() *cobra.Command {
	var (
		title       string
		description string
		steps       string
		obtained    string
		expected    string
		featureTag  string
		causeTag    string
		status      string
	)

	cmd := &cobra.Command{
		Use:   "update <id>",
		Short: "Update a bug",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			app := appctx.FromContext(cmd.Context())
			if err := requireAuth(app); err != nil {
				return app.Err(err)
			}

			bug := map[string]any{}
			if cmd.Flags().Changed("title") {
				bug["title"] = title
			}
			if cmd.Flags().Changed("description") {
				bug["description"] = description
			}
			if cmd.Flags().Changed("steps") {
				bug["steps_to_reproduce"] = steps
			}
			if cmd.Flags().Changed("obtained") {
				bug["obtained_result"] = obtained
			}
			if cmd.Flags().Changed("expected") {
				bug["expected_result"] = expected
			}
			if cmd.Flags().Changed("feature-tag") {
				bug["feature_tag"] = featureTag
			}
			if cmd.Flags().Changed("cause-tag") {
				bug["cause_tag"] = causeTag
			}
			if cmd.Flags().Changed("status") {
				bug["status"] = status
			}

			if len(bug) == 0 {
				return app.Err(output.ErrUsage("No update fields specified"))
			}

			body := map[string]any{"bug": bug}
			data, err := app.Client.Patch("/api/v1/bugs/"+args[0], body)
			if err != nil {
				return app.Err(err)
			}

			var resp struct {
				Bug any `json:"bug"`
			}
			if err := json.Unmarshal(data, &resp); err != nil {
				return app.Err(fmt.Errorf("failed to parse response: %w", err))
			}

			return app.OK(resp.Bug, output.WithSummary("Bug updated"))
		},
	}
	cmd.Flags().StringVar(&title, "title", "", "Bug title")
	cmd.Flags().StringVar(&description, "description", "", "Bug description")
	cmd.Flags().StringVar(&steps, "steps", "", "Steps to reproduce")
	cmd.Flags().StringVar(&obtained, "obtained", "", "Obtained result")
	cmd.Flags().StringVar(&expected, "expected", "", "Expected result")
	cmd.Flags().StringVar(&featureTag, "feature-tag", "", "Feature tag")
	cmd.Flags().StringVar(&causeTag, "cause-tag", "", "Root cause tag")
	cmd.Flags().StringVar(&status, "status", "", "Bug status")
	return cmd
}

func newBugsDeleteCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "delete <id>",
		Short: "Delete a bug",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			app := appctx.FromContext(cmd.Context())
			if err := requireAuth(app); err != nil {
				return app.Err(err)
			}

			if err := app.Client.Delete("/api/v1/bugs/" + args[0]); err != nil {
				return app.Err(err)
			}

			return app.OK(map[string]string{
				"deleted": args[0],
			}, output.WithSummary(fmt.Sprintf("Bug #%s deleted", args[0])))
		},
	}
}
